### PR TITLE
C++: new query for dead code after goto or break

### PR DESCRIPTION
**Pull request by rdmarsh2**

### CODE REVIEW

**Comment by rdmarsh2** [comment]
Tested on LGTM, A few resource leaks and minor logic bugs. The result on FFmpeg is a false positive that I think is because of an extractor bug.

https://lgtm.com/query/1506597157317/

File: cpp/ql/src/Critical/DeadCodeGoto.qhelp [hunk 1]
```diff
@@ -0,0 +1,28 @@
+<!DOCTYPE qhelp PUBLIC
+  "-//Semmle//qhelp//EN"
+  "qhelp.dtd">
+<qhelp>
+
+
+<overview>
+<p>
+Code following a goto or break statement will not be executed, unless there is a label or switch
```
Comment (semmledocs-ac) [inline cpp/ql/src/Critical/DeadCodeGoto.qhelp]: `Code following` > `Code immediately following`

**Comment by ian-semmle** [comment]
As discussed out-of-band, on ffmpeg the do-loop has been put inside a generated block, which also contains a generated label.

File: cpp/ql/src/Critical/DeadCodeGoto.ql [hunk 2]
```diff
@@ -0,0 +1,32 @@
+/**
+ * @name Dead code due to goto or break statement
```
Comment (geoffw0) [inline cpp/ql/src/Critical/DeadCodeGoto.ql]: As written the query will also catch `continue` statements that are followed by dead code, should we mention this anywhere?

File: cpp/ql/src/Critical/DeadCodeGoto.ql [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+}
+
+from JumpStmt js, Block b, int i, Stmt s
+where b.getStmt(i) = js
+  and s = getNextRealStmt(b, i)
+  // the next statement isn't jumped to
+  and not s instanceof LabelStmt
+  and not s instanceof SwitchCase
+  // the next statement isn't breaking out of a switch
+  and not s.(BreakStmt).getBreakable() instanceof SwitchStmt
```
Comment (geoffw0) [inline cpp/ql/src/Critical/DeadCodeGoto.ql]: Is this just to allow people to write:
```
  case VALUE:
  {
    goto somewhere_else;
  } break; // bit silly IMO, but hardly worth flagging?
```

File: cpp/ql/src/Critical/DeadCodeGoto.ql [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+  // the jump isn't a goto into the body of the next statement
+  and not exists (LabelStmt ls | s.(Loop).getStmt().getAChild*() = ls | ls.getName() = js.(GotoStmt).getName())
```
Comment (geoffw0) [inline cpp/ql/src/Critical/DeadCodeGoto.ql]: Why does it matter that this jump and the label match?
```
if (cond)
{
  goto target;
}
goto somewhere_else;
while (x < 10) // not dead code
{
target:
  x++;
}
```

File: cpp/ql/test/query-tests/Critical/DeadCodeGoto/test.cpp [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+		goto label; // GOOD
+	case 5:
+		goto label;; // GOOD
+	default:
+		x++;
+	}
+	label:
+	return x;
+}
+
```
Comment (geoffw0) [inline cpp/ql/test/query-tests/Critical/DeadCodeGoto/test.cpp]: Can we have a case along the lines of:
```
if (cond) {
  x++;
} else goto end; // GOOD
x++;
end:
```

File: cpp/ql/src/Critical/DeadCodeGoto.ql (hunk 4 above)
Comment (geoffw0) [reply cpp/ql/src/Critical/DeadCodeGoto.ql]: And if we want to consider really weird edge cases, the `LabelStmt` could be a `SwitchCase`:
```
switch (1)
{
  goto end;
  while (x < 10) // not dead code
  {
case 1:
    x++;
  } break;
}
end:
```

File: cpp/ql/src/Critical/DeadCodeGoto.ql (hunk 3 above)
Comment (rdmarsh2) [reply cpp/ql/src/Critical/DeadCodeGoto.ql]: Yes - some projects use that pattern very heavily

File: cpp/ql/src/Critical/DeadCodeGoto.ql (hunk 4 above)
Comment (rdmarsh2) [reply cpp/ql/src/Critical/DeadCodeGoto.ql]: That's a good point - I'll change that and add tests for both those cases

File: cpp/ql/test/query-tests/Critical/DeadCodeGoto/test.cpp (hunk 5 above)
Comment (rdmarsh2) [reply cpp/ql/test/query-tests/Critical/DeadCodeGoto/test.cpp]: added

**Pull request merged**